File: portfolio/projects/D_INSA/INSA/backend/app/schemas/comp_indicator.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class CompBehaviorBase(BaseModel):
    level: int = Field(ge=1, le=5)
    description: str | None = None


class CompBehaviorCreate(CompBehaviorBase):
    pass


class CompBehaviorResponse(CompBehaviorBase):
    id: int
    indicator_id: int

    model_config = {"from_attributes": True}


class CompIndicatorBase(BaseModel):
    year: int
    code: str
    name: str
    description: str | None = None
    weight: Decimal | None = None


class CompIndicatorCreate(CompIndicatorBase):
    behaviors: list[CompBehaviorCreate] = []


class CompIndicatorResponse(CompIndicatorBase):
    id: int
    created_at: datetime | None = None
    behaviors: list[CompBehaviorResponse] = []

    model_config = {"from_attributes": True}

File: portfolio/projects/D_INSA/INSA/frontend/src/api/compIndicator.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import apiClient from "./client";

export interface CompBehavior {
  id: number;
  indicator_id: number;
  level: number;
  description: string | null;
}

export interface CompIndicator {
  id: number;
  year: number;
  code: string;
  name: string;
  description: string | null;
  weight: string | null;
  created_at: string | null;
  behaviors: CompBehavior[];
}

export interface CompIndicatorCreatePayload {
  year: number;
  code: string;
  name: string;
  description?: string | null;
  weight?: string | number | null;
  behaviors?: { level: number; description?: string | null }[];
}

export function useCompIndicators(year: number | null) {
  return useQuery({
    queryKey: ["eval", "comp", "indicators", { year }],
    queryFn: async () => {
      const { data } = await apiClient.get<CompIndicator[]>(
        "/eval/comp/indicators",
        { params: { year: year ?? undefined } },
      );
      return data;
    },
  });
}

export function useCreateCompIndicator() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: CompIndicatorCreatePayload) => {
      const { data } = await apiClient.post<CompIndicator>(
        "/eval/comp/indicators",
        body,
      );
      return data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["eval", "comp", "indicators"] });
    },
  });
}
